### ISSUE TITLE
Verbose mode: provides command failure details

### ISSUE THREAD

**Issue by ael-code** [anishathalye/dotbot#104]
Even if I enable the verbose flag, no details are provided in case of command failure

**Comment by anishathalye** [comment]
Perhaps related to #73.

Would it make sense to assume `stdout: true` and `stderr: true` when given the verbose flag?

**Comment by ael-code** [comment]
I didn't read the code, but I think that on `stdout` and `stderr` you will only information outputted from the invoked command but what about the invoker errors? I mean 'command not found', 'permission denied' and so on...

**Comment by guaguanco** [comment]
The problem is that each class, including the plugins, creates its own Messenger object for logging  and the level coming from the arguments is set only in `dotbot.cli.main`. I guess that a possible solution could be to use the context class to propagate the options to all classes

**Comment by jesseleite** [comment]
>Would it make sense to assume stdout: true and stderr: true when given the verbose flag?

@anishathalye 100% this! I find myself manually running shell commands, or tweaking `stdout` / `stderr` settings when debugging something in my install file. Would be so much nicer to just `-v` verbose output and run the install file 👍

**Comment by anishathalye** [comment]
I could see it being useful in certain situations but harmful in others. Eg if you have a config file which runs a command that produces tons of useless output, and you want to debug something else in the file with —verbose, it could be annoying.

**Comment by jesseleite** [comment]
I get that, but having to open the yaml and save `stderr: true` to see why a command is failing is also a bit obtuse, if I want to see error output only once.

Maybe you could leave `--verbose`, and instead add `--with-output` / `--with-errors` flags to compliment the hardcoded `stdout` and `stderr` yaml settings? What are your thoughts on this? 🙂

**Comment by etkeys** [comment]
An alternative to adding more cli options, an idea I got from [FAKE](https://github.com/fsharp/FAKE/blob/5.20.3/src/app/Fake.netcore/Cli.fs#L34-L37) would be to change verbose current action to `action='count'`. With verbose as a counter, passing this to Dispatcher to force set context defaults for shell `stdout: true` and `stderr: true` when verbose is 2 or higher.

This would not change the current usage of verbose, except if someone was already passing `--vebose` more than once (is this a valid expectation to consider or is it "undefined behaviour"). Also, setting the log level would just need to be tweaked to expect `int` instead of `bool`.

Only down side is with force setting context defaults. Everytime `Context.set_defaults` is called, would need to follow up with reapplying the verbose level forced values.

**Comment by anishathalye** [comment]
How about making `--verbose` set `stdout: True` and `stderr: True` if it has not been explicitly set, but if there is a `stdout: False` etc., then `--verbose` won't override it?

**Comment by etkeys** [comment]
In an effort to gain some insight, I took a peek at @jesseleite `install.conf.yaml`. Their setup is similar to this
```yaml
- defaults:
    shell:
        stderr: false
        stdout: false

- shell:
    - description: clone a repo
      command: git clone https://github.com/anishathalye/dotbot.git
```
`defaults` specifies `stderr: false` and `stdout: false`. Even though these are the implicit default values for these parameters, no fault for being explicit. The objective is "I want to see the output of shell commands for just one run (even if I normally ask to not see it) in case I need to debug something and I don't want to modify/copy `install.conf.yaml` because <reasons>".

> How about making `--verbose` set `stdout: True` and `stderr: True` if it has not been explicitly set, but if there is a `stdout: False` etc., then `--verbose` won't override it?

1. Given the earlier example, since `defaults` explicitly sets `stderr` and `stdout`, there is no runtime change and problem is not solved.
1. Those that rely on the implicit default for those parameters could suddenly find a lot more unexpected noise through no fault of their own.

**Comment by jesseleite** [comment]
@anishathalye @etkeys Regarding this...
```
- defaults:
    shell:
        stderr: false
        stdout: false
```
I normally don't have that in my install.conf.yaml file. I know those are implicitly `false` by default, and only had them in there temporarily, as I just got a new machine and have been debugging errors. Pretty silly of me, but it's just easier to toggle `true` if the keys are already in the yaml; I have no other reason for explicitly setting those to `false` lol.

>The objective is "I want to see the output of shell commands for just one run (even if I normally ask to not see it) in case I need to debug something and I don't want to modify/copy install.conf.yaml because...reasons".

Exactly. In theory I should be able to just do this:
```
- defaults:
    shell:
        stderr: true
        stdout: false
```
This would hide successful output, only showing errors when they pop. However, not all commands output proper `stderr` output when they error. I've found myself in multiple failing situations where I need to see both `stderr` and `stdout` to figure out what's going on.

>How about making `--verbose` set `stdout: True` and `stderr: True` if it has not been explicitly set, but if there is a `stdout: False` etc., then `--verbose` won't override it?

If this is how you implement it, I would not complain! However, maybe that solution would be a bit too magical? My thought was that `--with-output` and `--with-errors` flags would be very simple and self-explanatory.

**Comment by etkeys** [comment]
I slept on this and thought about it more. Whether the direction is to change the under the hood mechanics of `--verbose` or change how it is implemented or even provide new cli flags, acting on this change should be pushed down to *shell* as much as possible.

I believe this issue is a subset of #249 where plugins need more access to runtime context values. If *shell* was aware that *dotbot* is running in verbose mode or that specific cli flags were used, *shell* could deal with when to override `stderr` and `stdout` that were set by the config file. This way, the core application doesn't have to manage updates to defaults and force overriding values from the config file. The core application just stores cli flags as usual and the plugin is responsible for altering its behavior accordingly.

**Comment by etkeys** [comment]
@anishathalye,
I did a thing. I have a [commit](https://github.com/etkeys/dotbot/commit/ffc1ee05f307c48e67b08153606ebef6c00890f3) that can resolve this issue. However, these changes depend on the changes in PR #250. I'm not sure how to create a PR into your repo that also indicates/forces a pending PR to be implemented first. I could merge these changes into those for #250 and bundle them, but that feels awkward to me.

Any advice for how you'd like to review these changes?

**Comment by jesseleite** [comment]
@etkeys nice! How would that work? By passing `--stderr` / `--stdout`?

**Comment by etkeys** [comment]
@jesseleite,
As of now, to enable this behavior one would need to pass `--verbose` two times. For example, `./install -vv`. I opted for this approach because it is backwards compatible and doesn't affect the code base too much. However, this is still subject to change because a final decision for how to enable this behavior hasn't been made.

**Comment by jesseleite** [comment]
Oh that's what you meant by the count on verbose, very cool.

**Comment by anishathalye** [comment]
> Any advice for how you'd like to review these changes?

You can send it as a separate PR that includes the commits from the other PR, and just mention as a reminder for me that it's dependent on another PR. I'll review/merge the dependency first.

**Comment by jesseleite** [comment]
This is awesome, thank you @anishathalye!
